accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,231 @@
+func (d *depTreeBuilder) BuildDepTrees() (roots []*depTreeNode, nonRoots []*depTreeNode) {
+	// before the trees of dependencies are known, consider that any provided contract could be a root.
+	for pattern, _ := range d.contracts {
+		d.roots[pattern] = struct{}{}
+	}
+	for pattern, contract := range d.contracts {
+		d.buildDepTrees(pattern, contract)
+	}
+	for pattern, _ := range d.contracts {
+		if _, ok := d.roots[pattern]; ok {
+			roots = append(roots, d.subtrees[pattern])
+		} else {
+			nonRoots = append(nonRoots, d.subtrees[pattern])
+		}
+	}
+	return roots, nonRoots
+}
+
+func newDepTreeBuilder(overrides map[string]common.Address, contracts map[string]string) *depTreeBuilder {
+	return &depTreeBuilder{
+		overrides: overrides,
+		contracts: contracts,
+		subtrees:  make(map[string]*depTreeNode),
+		roots:     make(map[string]struct{}),
+	}
+}
+
+// DeployFn deploys a contract given a deployer and optional input.  It returns
+// the address and a pending transaction, or an error if the deployment failed.
+type DeployFn func(input, deployer []byte) (common.Address, *types.Transaction, error)
+
+// depTreeDeployer is responsible for taking a dependency, deploying-and-linking its components in the proper
+// order.  A depTreeDeployer cannot be used after calling LinkAndDeploy other than to retrieve the deployment result.
+type depTreeDeployer struct {
+	deployedAddrs map[string]common.Address
+	deployerTxs   map[string]*types.Transaction
+	input         map[string][]byte // map of the root contract pattern to the constructor input (if there is any)
+	deploy        DeployFn
+}
+
+// linkAndDeploy recursively deploys a contract and its dependencies:  starting by linking/deploying its dependencies.
+// The deployment result (deploy addresses/txs or an error) is stored in the depTreeDeployer object.
+func (d *depTreeDeployer) linkAndDeploy(node *depTreeNode) error {
+	// don't deploy contracts specified as overrides.  don't deploy their dependencies.
+	if node.overrideAddr != nil {
+		return nil
+	}
+
+	// if this contract/library depends on other libraries deploy them (and their dependencies) first
+	for _, childNode := range node.children {
+		if err := d.linkAndDeploy(childNode); err != nil {
+			return err
+		}
+	}
+	// if we just deployed any prerequisite contracts, link their deployed addresses into the bytecode to produce
+	// a deployer bytecode for this contract.
+	deployerCode := node.unlinkedCode
+	for _, child := range node.children {
+		var linkAddr common.Address
+		if child.overrideAddr != nil {
+			linkAddr = *child.overrideAddr
+		} else {
+			linkAddr = d.deployedAddrs[child.pattern]
+		}
+		deployerCode = strings.ReplaceAll(deployerCode, "__$"+child.pattern+"$__", strings.ToLower(linkAddr.String()[2:]))
+	}
+
+	// Finally, deploy the contract.
+	addr, tx, err := d.deploy(d.input[node.pattern], common.Hex2Bytes(deployerCode))
```

Comment:
Whoa!

The `deploy` function is a 
```golang
// DeployFn deploys a contract given a deployer and optional input.  It returns
// the address and a pending transaction, or an error if the deployment failed.
type DeployFn func(input, deployer []byte) (common.Address, *types.Transaction, error)
```
Yet, you invoke it with the `deployer []byte` being hex-encoded bytes. If the input is `[]byte`, and the docs says nothing else, then I'd _definitely_ assume that one would _not_ have to hex-encode the input. 

Non-encoded is the canonical form, IMO the `DeployFn` should use that.

Reply:
oof.  yeah, I agree with this.

Reply:
Ah wait, My bad! This is `Hex2Bytes`, I read it the other way around. So you are not encoding to hex, you are decoding _from_ hex.

Reply:
But it's a bit dangerous, `Hex2Bytes` just gobbles up any error, returning an empty slice. You probably want to error hard on invalid hex-encoding...

Reply:
Wait no.  I think this is a false alarm.  The `deployerCode` here is a hex-string which I convert to non-hex-encoded bytes with `Hex2Bytes`.

Reply:
Second edit to last comment:  yeah, I don't see the issue here.